homeassistant/components/netgear/router.py

```diff
@@ -0,0 +1,209 @@
+
+
+class NetgearRouter:
+    """Representation of a Netgear router."""
+
+    def __init__(self, hass: HomeAssistantType, entry: ConfigEntry) -> None:
+        """Initialize a Netgear router."""
+        self.hass = hass
+        self.entry_id = entry.entry_id
+        self.unique_id = entry.unique_id
+        self._url = entry.data.get(CONF_URL)
+        self._host = entry.data.get(CONF_HOST)
+        self._port = entry.data.get(CONF_PORT)
+        self._ssl = entry.data.get(CONF_SSL)
+        self._username = entry.data.get(CONF_USERNAME)
+        self._password = entry.data[CONF_PASSWORD]
+
+        self._info = None
+        self.model = None
+        self.device_name = None
+        self.firmware_version = None
+
+        self._method_version = 1
+        consider_home_int = entry.options.get(
+            CONF_CONSIDER_HOME, DEFAULT_CONSIDER_HOME.total_seconds()
+        )
+        self._tracked_list = convert_tracked_list(
+            entry.options.get(CONF_TRACKED_LIST, "")
+        )
+        self._consider_home = timedelta(seconds=consider_home_int)
+
+        self._api: Netgear = None
+        self._attrs = {}
+
+        self.devices = {}
+
+        self.listeners = []
+        self._unsub_dispatcher = None
+
+    def _setup(self) -> None:
+        """Set up a Netgear router sync portion."""
+        self._api = get_api(
+            self._password,
+            self._host,
+            self._username,
+            self._port,
+            self._ssl,
+            self._url,
+        )
+
+        self._info = self._api.get_info()
+        self.device_name = self._info["DeviceName"]
+        self.model = self._info["ModelName"]
+        self.firmware_version = self._info["Firmwareversion"]
+
+        if self.model in MODELS_V2:
+            self._method_version = 2
+
+    async def async_setup(self) -> None:
+        """Set up a Netgear router."""
+        await self.hass.async_add_executor_job(self._setup)
+
+        # set already known devices to away instead of unavailable
+        entity_registry = er.async_get(self.hass)
+        entries = er.async_entries_for_config_entry(entity_registry, self.entry_id)
+        for entity_entry in entries:
+            self.devices[entity_entry.unique_id] = {
+                "mac": entity_entry.unique_id,
+                "name": entity_entry.original_name,
+                "active": False,
+                "last_seen": dt_util.utcnow() - timedelta(days=365),
+                "device_model": None,
+                "device_type": None,
+                "type": None,
+                "link_rate": None,
+                "signal": None,
+                "ip": None,
+            }
+
+        await self.async_update_device_trackers()
+        self._unsub_dispatcher = async_track_time_interval(
+            self.hass, self.async_update_device_trackers, SCAN_INTERVAL
+        )
+
+        async_dispatcher_send(self.hass, self.signal_device_new)
+
+    async def async_unload(self) -> None:
```

Comment:
Can this ever get called before the dispatcher is setup?

Reply:
I don't understand the exact workings.
If async_setup is done the dispatcher is setup (I think), and I presume that async_unload can only be called after async_setup is complete (I hope the core of HomeAssistant handels this?)

Reply:
Looks like its only ever called in `async_unload_entry` so it should be race safe.